planner/optimize.go

```diff
@@ -99,11 +99,34 @@ func GetExecuteForUpdateReadIS(node ast.Node, sctx sessionctx.Context) infoschem
 	return nil
 }
 
+func matchSQLBinding(node ast.Node, sctx sessionctx.Context) (bindRecord *bindinfo.BindRecord, scope string, matched bool) {
+	useBinding := sctx.GetSessionVars().UsePlanBaselines
+	stmtNode, ok := node.(ast.StmtNode)
+	if !ok {
+		useBinding = false
+	}
+	if !useBinding {
+		return nil, "", false
+	}
+	var err error
+	bindRecord, scope, err = getBindRecord(sctx, stmtNode)
+	if err != nil || bindRecord == nil || len(bindRecord.Bindings) == 0 {
+		return nil, "", false
+	}
+	return bindRecord, scope, true
+}
+
 // Optimize does optimization and creates a Plan.
 // The node must be prepared first.
 func Optimize(ctx context.Context, sctx sessionctx.Context, node ast.Node, is infoschema.InfoSchema) (plannercore.Plan, types.NameSlice, error) {
 	sessVars := sctx.GetSessionVars()
 
+	bindRecord, scope, useBinding := matchSQLBinding(node, sctx)
```

Comment:
Maybe we should put this part behind the [codes](https://github.com/pingcap/tidb/blob/fec2938c1379270bf9939822c1abfe3d7244c174/planner/optimize.go#L107-L128) for less change.